[Style] Merge animation/transition and background/mask list code

weinig opened the pull request:
#### de89ec12b9a340f3482065df31015cdc4b15b5e6
<pre>
[Style] Merge animation/transition and background/mask list code
<a href="https://bugs.webkit.org/show_bug.cgi?id=301581">https://bugs.webkit.org/show_bug.cgi?id=301581</a>

Reviewed by Darin Adler.

Unifies the implementations of background, mask, animation and transition
coordinated list values around Style::CoordinatedListValue. To make this
a bit less code to manually write, a set of macros have been added that
simplify adding unset/set/filled state for each property.

While mostly the same, background/mask and animation/transition differ in
what they return from getComputedStyle, with background/mask returning
values based on the filtered (used) values and background/mask returning
values based on the unfiltered (computed) values. To indicate this to the
extractor, the style types have a constexpr bool, `computedValueUsesUsedValues`,
that behavior is conditionalized on.

* Source/WebCore/Headers.cmake:
* Source/WebCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedLambdaCapturesCheckerExpectations:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/animation/WebAnimationUtilities.cpp:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/scripts/process-css-properties.py:
* Source/WebCore/page/ElementTargetingController.cpp:
* Source/WebCore/page/InteractionRegion.cpp:
* Source/WebCore/page/LocalFrameView.cpp:
* Source/WebCore/rendering/BackgroundPainter.cpp:
* Source/WebCore/rendering/InlineBoxPainter.cpp:
* Source/WebCore/rendering/ReferencedSVGResources.cpp:
* Source/WebCore/rendering/RenderBox.cpp:
* Source/WebCore/rendering/RenderElement.cpp:
* Source/WebCore/rendering/RenderImage.cpp:
* Source/WebCore/rendering/RenderInline.cpp:
* Source/WebCore/rendering/RenderLayer.cpp:
* Source/WebCore/rendering/RenderLayerBacking.cpp:
* Source/WebCore/rendering/RenderLayerCompositor.cpp:
* Source/WebCore/rendering/RenderLayerModelObject.cpp:
* Source/WebCore/rendering/RenderTableCell.cpp:
* Source/WebCore/rendering/RenderView.cpp:
* Source/WebCore/rendering/TextPaintStyle.cpp:
* Source/WebCore/rendering/ios/RenderThemeIOS.mm:
* Source/WebCore/rendering/style/RenderStyle.cpp:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleConstants.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
* Source/WebCore/rendering/style/StyleBackgroundData.h:
* Source/WebCore/rendering/style/StyleMiscNonInheritedData.h:
* Source/WebCore/rendering/svg/SVGRenderTreeAsText.cpp:
* Source/WebCore/rendering/svg/legacy/SVGResources.cpp:
* Source/WebCore/rendering/svg/legacy/SVGResourcesCache.cpp:
* Source/WebCore/style/StyleAdjuster.cpp:
* Source/WebCore/style/StyleBuilderConverter.h:
* Source/WebCore/style/StyleBuilderCustom.h:
* Source/WebCore/style/StyleExtractorConverter.h:
* Source/WebCore/style/StyleExtractorCustom.h:
* Source/WebCore/style/StyleExtractorSerializer.h:
* Source/WebCore/style/StyleInterpolationWrappers.h:
* Source/WebCore/style/StylePendingResources.cpp:
* Source/WebCore/style/Styleable.cpp:
* Source/WebCore/style/values/animations/StyleAnimation.cpp:
* Source/WebCore/style/values/animations/StyleAnimation.h:
* Source/WebCore/style/values/backgrounds/StyleBackgroundLayer.cpp:
* Source/WebCore/style/values/backgrounds/StyleBackgroundLayer.h:
* Source/WebCore/style/values/backgrounds/StyleBackgroundLayers.h: Added.
* Source/WebCore/style/values/backgrounds/StyleFillLayers.h:
* Source/WebCore/style/values/masking/StyleMaskLayer.cpp:
* Source/WebCore/style/values/masking/StyleMaskLayer.h:
* Source/WebCore/style/values/masking/StyleMaskLayers.h: Added.
* Source/WebCore/style/values/primitives/StyleCoordinatedValueList.h:
* Source/WebCore/style/values/primitives/StyleCoordinatedValueListValue.h: Added.
* Source/WebCore/style/values/transitions/StyleSingleTransitionProperty.h:
* Source/WebCore/style/values/transitions/StyleTransition.cpp:
* Source/WebCore/style/values/transitions/StyleTransition.h:

Canonical link: <a href="https://commits.webkit.org/302416@main">https://commits.webkit.org/302416@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9ba45c0d4fc8cb6a60271772902910f283cf1102

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/129032 "check-webkit-style (failure)") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/1290 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/39868 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/136412 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/80389 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/130903 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/1222 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/1168 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/98243 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/80389 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/131979 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/944 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/115583 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/78887 "Passed tests") | 
| [❌ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/128382 "webkitpy-tests (failure)") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/872 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/33698 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/79691 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/109312 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/34196 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/138886 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/1085 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/1057 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/106780 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/1142 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/111919 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/106606 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27138 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/896 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/30443 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/53590 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/1159 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/64502 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/993 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/1042 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/1086 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/css/scripts/process-css-properties.py:
```diff
@@ -3671,10 +3645,19 @@ def _generate_css_property_names_h_property_constants(self, *, to):
+        to.write(f"template<CSSPropertyID C> struct PropertyNameConstant {{")
+        with to.indent():
+            to.write(f"static constexpr auto value = C;")
+            to.write(f"constexpr bool operator==(const PropertyNameConstant<C>&) const = default;")
```
I think the `<C>` can be omitted here.

```suggestion
            to.write(f"constexpr bool operator==(const PropertyNameConstant&) const = default;")
```

darinadler commented on Source/WebCore/rendering/style/RenderStyle.cpp:
```diff
@@ -2440,50 +2440,39 @@ const AtomString& RenderStyle::hyphenString() const
-            break;
-        }
-    }
+    list.prepareForUse();
```
I’m a little surprised we are using a local variable here.
```suggestion
    ensureAnimations().prepareForUse();
```
Same question about the other 3 functions.

darinadler commented on Source/WebCore/style/values/animations/StyleAnimation.cpp:
Should we refactor this so we don’t have to list every data member? Maybe we need a structure that holds all this data that is not reference counted?

weinig replied:
I really need to make WTF::Box work for this use case. It would be much cleaner. Let me see if I can do that.

darinadler commented on Source/WebCore/style/values/animations/StyleAnimation.h:
```diff
@@ -142,51 +102,48 @@ struct Animation {
-            && (!isRangeEndSet() || isRangeEndFilled());
-    }
+    // Support for the `animation-range` shorthand.
+    static SingleAnimationRange initialRange() { return { initialRangeStart(), initialRangeEnd() }; }
```
Can/should this also be `constexpr`?

weinig replied:
Alas, initialRangeStart() and initialRangeEnd() cannot be constexpr due to some Style::LengthPercentages in there (those require calc(), which requires malloc).

darinadler commented on Source/WebCore/style/values/animations/StyleAnimation.h:
```diff
@@ -142,51 +102,48 @@ struct Animation {
-    static void fillUnsetProperties(auto& list);
+    // Used for animation composite order sorting needed when backing a CSSAnimation object.
+    // https://drafts.csswg.org/css-animations-2/#animation-composite-order
+    uint64_t sortingIdentity() const { return reinterpret_cast<uint64_t>(m_data.ptr()); }
```
Sorting based on pointer values has not turned out well in the past. I’m dismayed to find this here. It’s not new code, though!

weinig replied:
Hm, good point. The reason this is currently ok is that its value is not actually used for the sorting itself, but rather only for checking identity:

```
    auto& aBackingAnimation = a.backingStyleAnimation();
    auto& bBackingAnimation = b.backingStyleAnimation();
    for (auto& animation : cssAnimationList->usedValues()) {
        if (animation.sortingIdentity() == aBackingAnimation.sortingIdentity())
            return true;
        if (animation.sortingIdentity() == bBackingAnimation.sortingIdentity())
            return false;
    }
```

 Really, this should return a type that only supports operator==, since actually using its numeric value would be bad.

darinadler commented on Source/WebCore/style/values/backgrounds/StyleBackgroundLayer.cpp:
```diff
@@ -26,47 +26,51 @@
-BackgroundLayer::BackgroundLayer(CSS::Keyword::None)
-    : BackgroundLayer { ImageOrNone { CSS::Keyword::None { } } }
+BackgroundLayer::BackgroundLayer(CSS::Keyword::None keyword)
+    : BackgroundLayer { }
 {
+    setImage(ImageOrNone { keyword });
 }
```
Should we do this instead?
```suggestion
BackgroundLayer::BackgroundLayer(CSS::Keyword::None keyword)
    : BackgroundLayer { ImageOrNone { keyword } }
{
}
```

weinig replied:
Yes.

darinadler commented on Source/WebCore/style/values/backgrounds/StyleBackgroundLayer.cpp:
```diff
@@ -26,47 +26,51 @@
 
 BackgroundLayer::BackgroundLayer(RefPtr<StyleImage>&& image)
-    : BackgroundLayer { ImageOrNone { WTFMove(image) } }
+    : BackgroundLayer { }
 {
+    setImage(ImageOrNone { WTFMove(image) });
 }
```
Should we do this instead?
```suggestion
BackgroundLayer::BackgroundLayer(RefPtr<StyleImage>&& image)
    : BackgroundLayer { ImageOrNone { WTFMove(image) } }
{
}
```

darinadler commented on Source/WebCore/style/values/backgrounds/StyleBackgroundLayer.h:
```diff
@@ -67,27 +92,15 @@ struct BackgroundLayer {
+    static FillAttachment initialAttachment() { return FillAttachment::ScrollBackground; }
+    static FillBox initialClip() { return FillBox::BorderBox; }
+    static FillBox initialOrigin() { return FillBox::PaddingBox; }
+    static BlendMode initialBlendMode() { return BlendMode::Normal; }
```
Should some of these be `constexpr`.

weinig replied:
Yes.

darinadler commented on Source/WebCore/style/values/backgrounds/StyleBackgroundLayers.h:
```diff
@@ -0,0 +1,37 @@
+
+#include <WebCore/StyleBackgroundLayer.h>
+#include <WebCore/StyleCoordinatedValueList.h>
+#include <WebCore/StyleFillLayers.h>
```
Why include this header?

weinig replied:
It was needed in an earlier iteration. Not anymore though. Will remove.

darinadler commented on Source/WebCore/style/values/masking/StyleMaskLayer.cpp:
```diff
@@ -26,46 +26,51 @@
 
 MaskLayer::MaskLayer(RefPtr<StyleImage>&& image)
-    : MaskLayer { ImageOrNone { WTFMove(image) } }
+    : MaskLayer { }
 {
+    setImage(ImageOrNone { WTFMove(image) });
 }
```
Why do we need this change?

weinig replied:
It insures that m_imageState is set to “Set”, which is now needed even in the initial case.

darinadler commented on Source/WebCore/style/values/masking/StyleMaskLayer.cpp:
```diff
@@ -26,46 +26,51 @@
-    , m_composite(static_cast<unsigned>(MaskLayer::initialFillComposite()))
-    , m_maskMode(static_cast<unsigned>(MaskLayer::initialFillMaskMode()))
+    : MaskLayer { }
 {
+    setImage(WTFMove(image));
 }
 
-MaskLayer::MaskLayer(CSS::Keyword::None)
-    : MaskLayer { ImageOrNone { CSS::Keyword::None { } } }
+MaskLayer::MaskLayer(CSS::Keyword::None keyword)
+    : MaskLayer { }
 {
+    setImage(ImageOrNone { keyword });
 }
```
Why do we need this change?

darinadler commented on Source/WebCore/style/values/masking/StyleMaskLayers.h:
```diff
@@ -0,0 +1,37 @@
+#pragma once
+
+#include <WebCore/StyleCoordinatedValueList.h>
+#include <WebCore/StyleFillLayers.h>
```
Why do we need this include?

darinadler commented on Source/WebCore/style/values/masking/StyleMaskLayers.h:
```diff
@@ -0,0 +1,37 @@
+namespace WebCore {
+namespace Style {
+
+using MaskLayers = CoordinatedValueList<MaskLayer>;
```
Can this be done with forward declarations instead?

weinig replied:
It can. These headers are probably not necessary anymore. Will remove.

weinig replied:
Actually, it makes the includes and members consistent in the Style data classes, so I'm going to keep it. RenderStyle does do the forward declaration though.

darinadler commented on Source/WebCore/style/values/backgrounds/StyleBackgroundLayers.h:
```diff
@@ -0,0 +1,37 @@
+namespace WebCore {
+namespace Style {
+
+using BackgroundLayers = CoordinatedValueList<BackgroundLayer>;
```
Can this be done with forward declarations instead?

darinadler reviewed (approved):
Bold, I guess, to review the pull request before knowing the tests as passing.

weinig committed:
[Style] Merge animation/transition and background/mask list code
https://bugs.webkit.org/show_bug.cgi?id=301581

Reviewed by Darin Adler.

Unifies the implementations of background, mask, animation and transition
coordinated list values around Style::CoordinatedListValue. To make this
a bit less code to manually write, a set of macros have been added that
simplify adding unset/set/filled state for each property.

While mostly the same, background/mask and animation/transition differ in
what they return from getComputedStyle, with background/mask returning
values based on the filtered (used) values and background/mask returning
values based on the unfiltered (computed) values. To indicate this to the
extractor, the style types have a constexpr bool, `computedValueUsesUsedValues`,
that behavior is conditionalized on.

* Source/WebCore/Headers.cmake:
* Source/WebCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedLambdaCapturesCheckerExpectations:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/animation/WebAnimationUtilities.cpp:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/scripts/process-css-properties.py:
* Source/WebCore/page/ElementTargetingController.cpp:
* Source/WebCore/page/InteractionRegion.cpp:
* Source/WebCore/page/LocalFrameView.cpp:
* Source/WebCore/rendering/BackgroundPainter.cpp:
* Source/WebCore/rendering/InlineBoxPainter.cpp:
* Source/WebCore/rendering/ReferencedSVGResources.cpp:
* Source/WebCore/rendering/RenderBox.cpp:
* Source/WebCore/rendering/RenderElement.cpp:
* Source/WebCore/rendering/RenderImage.cpp:
* Source/WebCore/rendering/RenderInline.cpp:
* Source/WebCore/rendering/RenderLayer.cpp:
* Source/WebCore/rendering/RenderLayerBacking.cpp:
* Source/WebCore/rendering/RenderLayerCompositor.cpp:
* Source/WebCore/rendering/RenderLayerModelObject.cpp:
* Source/WebCore/rendering/RenderTableCell.cpp:
* Source/WebCore/rendering/RenderView.cpp:
* Source/WebCore/rendering/TextPaintStyle.cpp:
* Source/WebCore/rendering/ios/RenderThemeIOS.mm:
* Source/WebCore/rendering/style/RenderStyle.cpp:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleConstants.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
* Source/WebCore/rendering/style/StyleBackgroundData.h:
* Source/WebCore/rendering/style/StyleMiscNonInheritedData.h:
* Source/WebCore/rendering/svg/SVGRenderTreeAsText.cpp:
* Source/WebCore/rendering/svg/legacy/SVGResources.cpp:
* Source/WebCore/rendering/svg/legacy/SVGResourcesCache.cpp:
* Source/WebCore/style/StyleAdjuster.cpp:
* Source/WebCore/style/StyleBuilderConverter.h:
* Source/WebCore/style/StyleBuilderCustom.h:
* Source/WebCore/style/StyleExtractorConverter.h:
* Source/WebCore/style/StyleExtractorCustom.h:
* Source/WebCore/style/StyleExtractorSerializer.h:
* Source/WebCore/style/StyleInterpolationWrappers.h:
* Source/WebCore/style/StylePendingResources.cpp:
* Source/WebCore/style/Styleable.cpp:
* Source/WebCore/style/values/animations/StyleAnimation.cpp:
* Source/WebCore/style/values/animations/StyleAnimation.h:
* Source/WebCore/style/values/backgrounds/StyleBackgroundLayer.cpp:
* Source/WebCore/style/values/backgrounds/StyleBackgroundLayer.h:
* Source/WebCore/style/values/backgrounds/StyleBackgroundLayers.h: Added.
* Source/WebCore/style/values/backgrounds/StyleFillLayers.h:
* Source/WebCore/style/values/masking/StyleMaskLayer.cpp:
* Source/WebCore/style/values/masking/StyleMaskLayer.h:
* Source/WebCore/style/values/masking/StyleMaskLayers.h: Added.
* Source/WebCore/style/values/primitives/StyleCoordinatedValueList.h:
* Source/WebCore/style/values/primitives/StyleCoordinatedValueListValue.h: Added.
* Source/WebCore/style/values/transitions/StyleSingleTransitionProperty.h:
* Source/WebCore/style/values/transitions/StyleTransition.cpp:
* Source/WebCore/style/values/transitions/StyleTransition.h:

Canonical link: https://commits.webkit.org/302416@main
